### PR TITLE
feat: Add source and sink metrics

### PR DESCRIPTION
**Pull request by jachym-tousek-keboola**
Jira: https://keboola.atlassian.net/browse/PSGO-805

**Changes:**
- Refactored sink pipeline to return bytes written.
- Added stream metrics.

-----------

### CODE REVIEW

File: internal/pkg/service/stream/sink/router/router.go
Comment (Matovidlo) [reply internal/pkg/service/stream/sink/router/router.go:83]: I suggest to create the metric, run the service, run for example k6 test and inspect the metrics on running instance.

File: internal/pkg/service/common/task/node_test.go
Comment (Matovidlo) [inline internal/pkg/service/common/task/node_test.go:653]: Why `unexpectedError` = `insufficientStorage`?

File: internal/pkg/service/stream/sink/router/router.go
Comment (Matovidlo) [inline internal/pkg/service/stream/sink/router/router.go:52]: Is it meters? Isn't this rather `metrics`?

File: internal/pkg/service/stream/sink/router/router.go
Comment (Matovidlo) [reply internal/pkg/service/stream/sink/router/router.go:276]: Basically the `error` should be handled in higher layer. If we find some problem when reading `Request.Body()` there is simply issue in the request and we send back the user error which should be reported in the layer above. 
Then simply report the `len(bytes)` and suppress the error

File: internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go
Comment (Matovidlo) [inline internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go:15]: I am not huge fan of more than `2` return values. Can we put the number into `RecordStatus`?

File: internal/pkg/service/stream/storage/level/local/encoding/encoder/csv/fastcsv/writer.go
Comment (Matovidlo) [inline internal/pkg/service/stream/storage/level/local/encoding/encoder/csv/fastcsv/writer.go:28]:
```suggestion
	for index, col := range *cols {
```
It's 2 numbers there and it would look better.

File: internal/pkg/service/stream/storage/level/local/encoding/encoder/csv/fastcsv/writer.go
Comment (Matovidlo) [inline internal/pkg/service/stream/storage/level/local/encoding/encoder/csv/fastcsv/writer.go:71]: Do we care about `n` when we have `error` while writing? E.g I wrote 1MB of characters but end up with error. The meter does not matter because there will be issue that it cannot fit the `row` anymore.

File: internal/pkg/telemetry/meter.go
Comment (Matovidlo) [inline internal/pkg/telemetry/meter.go:11]: Why did we change the naming? I would inderstand that if we have `IntCounter` and `FloatCounter` but we have just 1.
For Histogram it makes sense.

File: internal/pkg/service/common/task/node_test.go
Comment (michaljurecko) [reply internal/pkg/service/common/task/node_test.go:653]: Because we added support for  `errors.WithName` interface.

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [reply internal/pkg/service/stream/sink/router/router.go:83]: Probably forgot to mention that the histogram makes something like a separate metric for each interval.

-----------

For bytes should be `Counter` sufficient, we will know min/max/average (per project, sink, ...)

-----------

Default histogram buckets are:
```
.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10 (seconds)
```

------

You should set some custom buckets for our durations - so we can measure more precisely which bucket the duration falls into:

```
WithExplicitBucketBoundaries(.005, .01, .025, .05, .1, .15, .2,.25, .3,.4,.5,.75,1, 2.5, 5, 7.5, 10,20,30),
```

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [inline internal/pkg/service/stream/sink/router/router.go:85]: :thinking: 

`keboola.go.stream.source.in.duration` -> `keboola.go.stream.source.{x/rec/record}.duration`

`keboola.go.stream.sink.in.duration` -> `keboola.go.stream.sink.{x/rec/record}.duration`
`keboola.go.stream.sink.in.bytes` -> `keboola.go.stream.sink.{out/encoded/uncompressed/raw}.bytes`

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [reply internal/pkg/service/stream/sink/router/router.go:264]: Sure :)

-----------

BTW, I suggest to replace it with:
```
finalizationCtx := context.WithoutCancel(c.Ctx())
```

@Matovidlo  and I recently found out that context timeout in hot path requires 20% more memory allocation.

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [reply internal/pkg/service/stream/sink/router/router.go:276]: I suggest to add `BodyLen()`  method without error ( returns `0` on error).

The `BodyBytes()` method is caching body AND err, so it is OK.

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [inline internal/pkg/service/stream/sink/router/router.go:304]: Same, please don't use context timeout.

File: internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go
Comment (michaljurecko) [reply internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go:15]: RecordStatus here tells whether the record was `accepted` or `processed`.

I take it as a standard Write method + status, I don't mind.

File: internal/pkg/service/stream/storage/level/local/encoding/encoder/csv/fastcsv/writer.go
Comment (michaljurecko) [reply internal/pkg/service/stream/storage/level/local/encoding/encoder/csv/fastcsv/writer.go:71]: On error, the buffer is discarded, either the entire row is written or nothing, so it should return `0`.

File: internal/pkg/telemetry/meter.go
Comment (michaljurecko) [reply internal/pkg/telemetry/meter.go:11]: In my opinion, it is a good step for the future.

**Review by michaljurecko — APPROVED** [review]
LGTM

Try to think about beter metric names.

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [reply internal/pkg/service/stream/sink/router/router.go]: We probably just haven't finished it, there should definitely be 2.

File: internal/pkg/service/stream/storage/level/local/encoding/encoder/csv/fastcsv/writer.go
Comment (Matovidlo) [reply internal/pkg/service/stream/storage/level/local/encoding/encoder/csv/fastcsv/writer.go:71]: So basically it would be better to return `0, err`.

File: internal/pkg/telemetry/meter.go
Comment (Matovidlo) [reply internal/pkg/telemetry/meter.go:11]: Then it would be good to add other methods as well, it's just 1 line of code as preparation.

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [reply internal/pkg/service/stream/sink/router/router.go:264]: (timeout is not expected here, metrics are collected to the memory)

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [reply internal/pkg/service/stream/sink/router/router.go:52]: OpenTelemetry is using `meter`/`meters` wording, see GO pkgs.

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [reply internal/pkg/service/stream/sink/router/router.go:264]: - You should remove ``context.WithTimeout` , because performance.
- It is safe to remove, timeout is not expected.

File: internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go
Comment (michaljurecko) [reply internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go:15]: For me, these two solutions make sense:
- `(pipeline.RecordStatus, int, error)`
- `(pipeline.WriteResult, error)`, `pipeline.WriteResult`  is struct

Comment (michaljurecko) [reply internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go:15]: This is really a detail, I suggest you to call about it after Martin's vacation, and don't block the merge, it can always be adjusted.

Comment (Matovidlo) [reply internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go:15]: I would go with `WriteResult` seems better

File: internal/pkg/service/stream/sink/router/router.go
Comment (michaljurecko) [reply internal/pkg/service/stream/sink/router/router.go:52]: Ok, but this is official OTEL Go package:
image

Of course, feel free to rename it.

**Comment by Matovidlo** [comment]
@michaljurecko I do not have special check for metrics but it would be fine to have session how to check new meters/histograms before releasing to production. I used to use `prometheus` exporter docker image or `grafana` docker image and import source container to show the metrics in graph.
What do you suggest when inspecting metrics in dev environment

**Review by Matovidlo — APPROVED** [review]

**Comment by michaljurecko** [comment]
@Matovidlo it's a standard OpenTelemetry setup, in the dev/prod environment you can set up any environment you want, I don't know what the added value of that call should be from my side (I just want to use the time for more specific things).

**Review by michaljurecko — APPROVED** [review]

**Pull request merged**
